### DISCUSSION TITLE
[next13] - use client required to index exportation

### DISCUSSION THREAD

**Discussion by vine1993** [vercel/next.js#46259] [Help]
Hi, I have a folder called **components** under the src folder, outside the app folder.

All components there follow the same pattern.
```
// index.tsx 
//just re-exporting from the other files.
export * from './componentName'
```
```
componentName.tsx //this component its client side.
componentName.stories.tsx
types.ts
```

When I try to use this component on the page, with a server-side component as children, I got an error saying my component its not exported, even my component using` 'use client'` at the top.

What I figure out its that my index.tsx also needs to use ` 'use client'`  at the top, i just didn't understand why.

Can you guys help me how to figure out that?

**Comment by icyJoseph** [comment]
I just saw this in someone else's repository. I had to point directly at the component. 

Now sure if the docs say anything about this. I guess the barrel file causes some loss of directive? Weird. 

Without the `use client` at the top level of the barrel file (the one that just exports * stuff, React claimed the component was `undefined`, just like in your case.

I think it is best to make a small reproduction demo, and create an issue. This doesn't look like a mistake, but rather a bug, though I am not sure if the docs say anything specifically about barrel files.

**Reply by nareshbhatia** [reply]
I have been using barrel files to expose only the elements that I want to from a directory. However, lately this has caused more trouble than it is worth (like in this case). Also barreling seems to have issues with tree shaking. Ready to abandon it altogether!

What is interesting though is that in my case `import { Header } from '~/components'` did not work but `import { Header } from '~/components/Header'` worked - even though the latter had an `index.ts` file with no "use client":

```ts
// ~/components/Header/index.ts
export * from './Header';
```

**Comment by icyJoseph** [comment]
Alright, looks like a fix might be coming soon: https://github.com/vercel/next.js/issues/46293 issue opened by a Vercel team member.
